File: slice-vtu-results-centerlines/python/graphics.py
#!/usr/bin/env python

from os import path
from math import sqrt 

import vtk
print(" vtk version %s\n" % str(vtk.VTK_MAJOR_VERSION))

class MouseInteractorStyle(vtk.vtkInteractorStyleTrackballCamera):
    def __init__(self, pick_geometry=None, picking_keys=None, event_table=None):
        #self.AddObserver("LeftButtonPressEvent", self.leftButtonPressEvent)
        self.AddObserver("KeyPressEvent", self.onKeyPressEvent)
        self.AddObserver("CharEvent", self.onCharEvent)
        self.graphics = None
        self.pick_geometry = pick_geometry
        self.event_table = event_table
        self.selected_point = None
        self.selected_node_id = None
        self.selected_cell_id = None
        self.picking_keys = picking_keys 
        self.picked_actor = None

    def select_event(self, key, obj, event):
        '''Process a select event. 
        '''
        clickPos = self.GetInteractor().GetEventPosition()
        picker = vtk.vtkPropPicker()
        # picker = vtk.vtkCellPicker()
        #picker.SetTolerance(0.0005)
        #picker.SetTolerance(1.0)
        #picker.Pick(clickPos[0], clickPos[1], 0, self.GetDefaultRenderer())
        picker.Pick(clickPos[0], clickPos[1], 0, self.graphics.renderer)

        # Get selected geometry. 
        self.picked_actor = picker.GetActor()
        if self.picked_actor == None:
            #print("[select_event] No picked actor.")
            #self.OnLeftButtonDown()
            return
        position = picker.GetPickPosition()

        #cell_id = picker.GetCellId()
        #if cell_id == -1: 
        #    return

        print("select_event] ")
        print("select_event] Picked position: {0:g} {1:g} {2:g} ".format(position[0], position[1], position[2]))
        #print("select_event] Cell id is: {0:d}".format(cell_id))

        min_i = -1
        min_d = 1e5
        min_p = 3*[0.0]
        pick_geometry = self.pick_geometry
        points = pick_geometry.GetPoints()

        min_d = 1e6
        for i in range(points.GetNumberOfPoints()):
            p = 3*[0.0]
            points.GetPoint(i,p)
            dx = p[0] - position[0]
            dy = p[1] - position[1]
            dz = p[2] - position[2]
            d = sqrt(dx*dx + dy*dy + dz*dz)
            if d < min_d:
                min_d = d
                min_p[0] = p[0]
                min_p[1] = p[1]
                min_p[2] = p[2]
                min_i = i  

        print("[select_event] Picked node: {0:d} {1:g} {2:g} {3:g} ".format(min_i, min_p[0], min_p[1], min_p[2]))
        self.selected_node_id = min_i

        self.graphics.add_sphere(position, [1,1,0], radius=0.2, wire=True)
        #self.graphics.add_sphere(min_cell_pt1 , [1,1,0], radius=0.2, wire=True)
        #self.graphics.add_sphere(min_cell_pt2 , [1,0,0], radius=0.2, wire=False)

        ## Show picked point.
        points = vtk.vtkPoints()
        points.SetNumberOfPoints(1)
        #points.SetPoint(0, cell_pick_pos[0], cell_pick_pos[1], cell_pick_pos[2])
        points.SetPoint(0, min_p[0], min_p[1], min_p[2])
        #points.SetPoint(0, position[0], position[1], position[2])
        geom = vtk.vtkPolyData()
        geom.SetPoints(points)
        glyphFilter = vtk.vtkVertexGlyphFilter()
        glyphFilter.SetInputData(geom)
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputConnection(glyphFilter.GetOutputPort())
        actor = vtk.vtkActor()
        if key == 's':
            actor.GetProperty().SetColor(0.0, 1.0, 0.0)
        elif key == 't':
            actor.GetProperty().SetColor(1.0, 0.0, 0.0)
        else:
            actor.GetProperty().SetColor(1.0, 1.0, 1.0)
        actor.GetProperty().SetPointSize(10)
        actor.SetMapper(mapper)
        self.graphics.renderer.AddActor(actor)

        #self.window.Render()

    def onKeyPressEvent(self, renderer, event):
        '''Process a key press event.
        '''
        key = self.GetInteractor().GetKeySym()

        if key == 'u' and len(self.method_queue) != 0:
            last_method = self.method_queue.pop()
            last_method(undo=True) 
            last_actor = self.pick_glyph_actor_queue.pop()
            last_actor.SetVisibility(False)

        if (key in self.picking_keys):
            self.select_event(key, None, event)

        if self.event_table == None:
            return

        if (key in self.event_table):
            value = self.event_table[key]
            if type(value) == tuple:
                method = value[0]
                data = value[1]
            else:
                method = value
                data = None

            # Call the method with selected node/cell and any data.
            method(pick_geometry=self.pick_geometry, node_id=self.selected_node_id, cell_id=self.selected_cell_id, data=data, 
                   graphics=self.graphics)

        self.graphics.window.Render()

    def onCharEvent(self, renderer, event):
        '''Process an on char event.

        This is used to prevent passing the shortcut key 'w' to vtk which we use
        to write selected results and vtk uses to switch to wireframe display. 
        '''
        key = self.GetInteractor().GetKeySym()
        if (key != 'w'):
            self.OnChar()

class Graphics(object):

    def __init__(self, width=1000, height=1000):
        self.window = None
        self.renderer = None
        self.interactor = None
        self.pick_face = None
        self.face_actors = {}
        self.initialize_graphics(width, height)

    def init_picking(self, pick_geometry, picking_keys, event_table=None):

        # Create a trackball interacter to transoform the geometry using the mouse.
        interactor = vtk.vtkRenderWindowInteractor()
        interactor.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
        interactor.SetRenderWindow(self.window)
        self.interactor = interactor 

        # Add the custom style.
        style = MouseInteractorStyle(pick_geometry, picking_keys, event_table)
        style.graphics = self
        interactor.SetInteractorStyle(style)
        style.SetCurrentRenderer(self.renderer)

    def create_graphics_geometry(self, poly_data, sphere=False):
        '''Create geometry for display.
        '''
        mapper = vtk.vtkPolyDataMapper()
        if sphere:
            mapper.SetInputConnection(poly_data)
        else:
            mapper.SetInputData(poly_data)
        mapper.ScalarVisibilityOff()
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        return actor

    def initialize_graphics(self, width=500, height=500):
        ''' Create renderer and graphics window.
        '''
        self.renderer = vtk.vtkRenderer()
        self.window = vtk.vtkRenderWindow()
        self.window.AddRenderer(self.renderer)
        self.renderer.SetBackground(0.8, 0.8, 0.8)
        self.window.SetSize(width, height)

        # Create a trackball interacter to transoform the geometry using the mouse.
        self.interactor = vtk.vtkRenderWindowInteractor()
        self.interactor.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
        self.interactor.SetRenderWindow(self.window)

        # Add the custom style.
        '''
        style = MouseInteractorStyle(self)
        style.renderer = self.renderer
        style.graphics = self
        self.interactor.SetInteractorStyle(style)
        style.SetCurrentRenderer(self.renderer)
        '''

    def add_sphere(self, center, color, radius=1.0, wire=False):
        sphere = vtk.vtkSphereSource()
        sphere.SetCenter(center[0], center[1], center[2])
        sphere.SetRadius(radius)
        sphere.SetPhiResolution(16)
        sphere.SetThetaResolution(16)
        sphere.Update()

        polydata = sphere.GetOutput()
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputData(polydata)
        mapper.ScalarVisibilityOff()
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        if wire:
            actor.GetProperty().SetRepresentationToWireframe()
        actor.GetProperty().SetColor(color)
        actor.GetProperty().SetOpacity(0.5)
        self.renderer.AddActor(actor)

    def add_line(self, pt1, pt2, color=[1.0, 1.0, 1.0], width=2):
        line = vtk.vtkLineSource()
        line.SetPoint1(pt1);
        line.SetPoint2(pt2)
        line.Update()
        polydata = line.GetOutput()
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputData(polydata)
        mapper.ScalarVisibilityOff()
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        actor.GetProperty().SetColor(color[0], color[1], color[2])
        actor.GetProperty().SetLineWidth(width)
        self.renderer.AddActor(actor)

    def add_geometry(self, poly_data, color, wire=False, edges=False, sphere=False, line_width=1.0):
        gr_geom = self.create_graphics_geometry(poly_data, sphere)
        gr_geom.GetProperty().SetColor(color[0], color[1], color[2])

        if wire:
            gr_geom.GetProperty().SetRepresentationToWireframe()
            gr_geom.GetProperty().SetLineWidth(1.0)
        else:
            gr_geom.GetProperty().SetLineWidth(line_width)

        if edges:
            gr_geom.GetProperty().EdgeVisibilityOn();

        #gr_geom.GetProperty().SetRepresentationToWireframe()
        #gr_geom.GetProperty().SetEdgeColor(0.0, 0.0, 0.0)
        #gr_geom.GetProperty().EdgeVisibilityOn()
        self.renderer.AddActor(gr_geom)
        self.window.Render()
        self.window.SetWindowName("Explore Model")
        return gr_geom 

    def show(self):
        self.interactor.Start()

File: slice-vtu-results-centerlines/c++/Graphics.cpp

#include "Graphics.h" 

#include <vtkCellData.h>
#include <vtkCellPicker.h>
//
#include <vtkExtractSelectedPolyDataIds.h>
#include <vtkExtractSelection.h>
//
#include <vtkGenericCell.h>
#include <vtkLineSource.h>
//
#include <vtkNamedColors.h>
#include <vtkPointData.h>
//
#include <vtkSelection.h>
#include <vtkSelectionNode.h>
//
#include <vtkUnstructuredGrid.h>

//----------
// Graphics
//----------
//
Graphics::Graphics()
{
  mesh_ = nullptr;
  centerlines_ = nullptr;

  // Add a renderer.
  renderer_ = vtkSmartPointer<vtkRenderer>::New();
  renderer_->SetBackground(1.0, 1.0, 1.0);

  // Add a render window.
  render_window_ = vtkSmartPointer<vtkRenderWindow>::New();
  render_window_->AddRenderer(renderer_);
  render_window_->SetSize(1000, 1000);
  render_window_->Render();
  render_window_->SetWindowName("Simulation Results Surface Slicer");

  // Add window interactor to use trackball and intercept key presses.
  //
  render_window_interactor_ = vtkSmartPointer<vtkRenderWindowInteractor>::New();
  render_window_interactor_->SetRenderWindow(render_window_);

  interaction_style_ = vtkSmartPointer<MouseCenterlineInteractorStyle>::New();
  //interaction_style_ = vtkSmartPointer<MouseMeshInteractorStyle>::New();

  interaction_style_->set_graphics(this);
  render_window_interactor_->SetInteractorStyle(interaction_style_);
  interaction_style_->SetDefaultRenderer(renderer_);
  render_window_interactor_->SetInteractorStyle(interaction_style_);
}

Mesh* Graphics::get_mesh()
{
  return mesh_;
}
void Graphics::set_mesh(Mesh* mesh)
{
  mesh_ = mesh;
}

void Graphics::set_centerlines(Centerlines* centerlines)
{
  centerlines_ = centerlines;
  interaction_style_->set_centerlines(centerlines);
}

void Graphics::start()
{
  //render_window_interactor->Initialize();
  render_window_interactor_->Start();
}

//----------------
// CreateGeometry 
//----------------
//
vtkActor* Graphics::create_geometry(vtkPolyData* polyData)
{
  auto mapper = vtkDataSetMapper::New();
  mapper->SetInputData(polyData);
  mapper->ScalarVisibilityOff();
  auto actor = vtkActor::New();
  actor->SetMapper(mapper);
  return actor; 
}

//-------------
// AddGeometry
//-------------
//
void Graphics::add_geometry(vtkActor* geom)
{
  renderer_->AddActor(geom);
}

std::string Graphics::get_data_name() 
{
  return data_name_;
}

void Graphics::set_data_name(const std::string& name) 
{
/*
  if (mesh_ != nullptr) {
    if (mesh_->HasData(name)) {
      data_name_ = name;
    } else {
      std::cout << "WARNING: The mesh does not have data named '" << name << "'" << std::endl;
    }
  } else {
      std::cout << "WARNING: The mesh is not set for graphics." << "'" << std::endl;
  }
*/
}

// Create a circle
vtkActor* Graphics::create_circle()
{
  vtkSmartPointer<vtkRegularPolygonSource> polygonSource = vtkSmartPointer<vtkRegularPolygonSource>::New();
  polygonSource->SetNumberOfSides(50);
  polygonSource->SetRadius(5);
  polygonSource->SetCenter(0, 0, 0);

  vtkSmartPointer<vtkPolyDataMapper> mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
  mapper->SetInputConnection(polygonSource->GetOutputPort());;
      
  auto actor = vtkActor::New();
  actor->SetMapper(mapper);
  return actor;
}   

void Graphics::refresh()
{
  render_window_->Render();
}

/////////////////////////////////////////////////////////////////////////
//            M o u s e M e s h I n t e r a c t o r S t y l e          //
/////////////////////////////////////////////////////////////////////////

// This class defines a mouse interactor used to select a surface
// element (triangle) and display its nodal data.

MouseMeshInteractorStyle::MouseMeshInteractorStyle()
{
  selected_mapper_ = vtkSmartPointer<vtkDataSetMapper>::New();
  selected_actor_ = vtkSmartPointer<vtkActor>::New();
  colors_ = vtkSmartPointer<vtkNamedColors>::New();
}

void MouseMeshInteractorStyle::set_graphics(Graphics* graphics)
{
  graphics_ = graphics;
}

// Process a keyboard press event.
//
void MouseMeshInteractorStyle::OnKeyPress() 
{
  // Get the keypress.
  vtkRenderWindowInteractor *rwi = this->Interactor;
  std::string key = rwi->GetKeySym();
  std::cout << "[MouseMeshInteractorStyle::OnKeyPress] Pressed key: " << key << std::endl;

  if ((key == "Escape") || (key == "q")) {
    exit(0);
  }
  else if (key == "s") {
    select_cell();
  }

  // Forward events
  //vtkInteractorStyleTrackballCamera::OnLeftButtonDown();
}

//-------------
// select_cell 
//-------------
//
void MouseMeshInteractorStyle::select_cell() 
{
  std::cout << "[select_cell] " << std::endl;
  vtkSmartPointer<vtkNamedColors> colors = vtkSmartPointer<vtkNamedColors>::New();

  // Get the location of the click (in window coordinates)
  int* pos = this->GetInteractor()->GetEventPosition();

  vtkSmartPointer<vtkCellPicker> picker = vtkSmartPointer<vtkCellPicker>::New();
  picker->SetTolerance(0.1);

  // Pick from this location.
  picker->Pick(pos[0], pos[1], 0, this->GetDefaultRenderer());

  double* worldPosition = picker->GetPickPosition();
  auto cellID = picker->GetCellId();

  if (cellID == -1) {
    return;
  }

  std::cout << "Pick position is: " << worldPosition[0] << " " << worldPosition[1] << " " << worldPosition[2] << endl;
  select_mesh(cellID);
}

//-------------
// SelectMesh 
//-------------
//
void MouseMeshInteractorStyle::select_mesh(int cellID)
{ 
  std::cout << "[select_mesh] " << std::endl;
  vtkSmartPointer<vtkIdTypeArray> ids = vtkSmartPointer<vtkIdTypeArray>::New();
  ids->SetNumberOfComponents(1);
  ids->InsertNextValue(cellID);

  vtkSmartPointer<vtkSelectionNode> selectionNode = vtkSmartPointer<vtkSelectionNode>::New();
  selectionNode->SetFieldType(vtkSelectionNode::CELL);
  selectionNode->SetContentType(vtkSelectionNode::INDICES);
  selectionNode->SetSelectionList(ids);

  vtkSmartPointer<vtkSelection> selection = vtkSmartPointer<vtkSelection>::New();
  selection->AddNode(selectionNode);

  //if (graphics_->get_mesh()->IsSurface()) {
    //select_mesh(cellID, selection);
  //}

}

//-------------------
// SelectSurfaceMesh
//-------------------
//
/*
void MouseMeshInteractorStyle::select_mesh(int cellID, vtkSmartPointer<vtkSelection> selection)
{ 
  auto mesh = graphics_->GetMesh();
  vtkPolyData* polydata = vtkPolyData::SafeDownCast(mesh->GetMesh());
  auto dataName = graphics_->get_data_name();
  auto data = mesh->GetDataArray(dataName);

  // Extract a list of cells from polydata.
  //
  vtkSmartPointer<vtkExtractSelection> extractSelection = vtkSmartPointer<vtkExtractSelection>::New();
  //vtkSmartPointer<vtkExtractSelectedPolyDataIds> extractSelection = vtkSmartPointer<vtkExtractSelectedPolyDataIds>::New();
  extractSelection->SetInputData(0, polydata); 
  extractSelection->SetInputData(1, selection);
  extractSelection->Update();

  // Get selected cells as a vtkPolyData.
  //
  vtkSmartPointer<vtkUnstructuredGrid> selected = vtkSmartPointer<vtkUnstructuredGrid>::New();
  //vtkSmartPointer<vtkPolyData> selected = vtkSmartPointer<vtkPolyData>::New();
  selected->ShallowCopy(extractSelection->GetOutput());
  std::cout << "There are " << selected->GetNumberOfPoints() << " points in the selection." << std::endl;
  std::cout << "There are " << selected->GetNumberOfCells() << " cells in the selection." << std::endl;

   // Print cell ID and connectivity.
   //
   vtkGenericCell* cell = vtkGenericCell::New();
   polydata->GetCell(cellID, cell);
   auto numPts = cell->GetNumberOfPoints();
   auto nodeIDs = vtkIntArray::SafeDownCast(polydata->GetPointData()->GetArray("GlobalNodeID"));
   auto elemIDs = vtkIntArray::SafeDownCast(polydata->GetCellData()->GetArray("GlobalElementID"));
   auto elemID = elemIDs->GetValue(cellID);
   std::cout << "Cell ID is: " << cellID << std::endl;
   std::cout << "  Elem ID: " << elemID << std::endl;
   std::cout << "  Number of cell points: " << numPts << std::endl;
   std::cout << "  Connectivity: ";
   for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
     auto id = cell->PointIds->GetId(pointInd);
     auto nodeID = nodeIDs->GetValue(id);
     std::cout << nodeID << " ";
   }
   std::cout << std::endl;

   // Print cell data.
   if (dataName != "") {
     std::cout << std::endl;
     std::cout << "Cell Data: " << std::endl;
     std::cout << "  Name: " << dataName << std::endl;
     auto numComp = data->GetNumberOfComponents();

     for (vtkIdType pointInd = 0; pointInd < numPts; ++pointInd) {
       auto id = cell->PointIds->GetId(pointInd);
       auto nodeID = nodeIDs->GetValue(id);
       if (numComp == 1) { 
         double value = data->GetValue(id);
         std::cout << "  ID: " << nodeID << "   Value: " << value << std::endl;
       } else if (numComp == 3) { 
         auto v1 = data->GetComponent(id, 0);
         auto v2 = data->GetComponent(id, 1);
         auto v3 = data->GetComponent(id, 2);
         std::cout << "  Node ID: " << nodeID << "   Values: " << v1 << " " << v2 << " " << v3 << std::endl;
       }
     }
   }

  // Display the selected cell.
  //
  selected_mapper_->SetInputData(selected);
  selected_mapper_->ScalarVisibilityOff();
  selected_actor_->SetMapper(selected_mapper_);
  //selectedActor->GetProperty()->EdgeVisibilityOn();
  //selected_actor_->GetProperty()->SetRepresentationToWireframe();
  selected_actor_->GetProperty()->SetColor(colors_->GetColor3d("Green").GetData());
  selected_actor_->GetProperty()->SetLineWidth(4);

  this->Interactor->GetRenderWindow()->GetRenderers()->GetFirstRenderer()->AddActor(selected_actor_);
  this->Interactor->GetRenderWindow()->Render();
}
*/

//void MouseInteractorStyle::OnLeftButtonDown()
//{
//}

vtkStandardNewMacro(MouseMeshInteractorStyle);

/////////////////////////////////////////////////////////////////////////////////////
//            M o u s e C e n t e r l i n e I n t e r a c t o r S t y l e          //
/////////////////////////////////////////////////////////////////////////////////////

// This class defines a mouse interactor used to create a slice at a selected
// point on a surface centerline.

MouseCenterlineInteractorStyle::MouseCenterlineInteractorStyle()
{
  startSelected = false;
}

void MouseCenterlineInteractorStyle::set_centerlines(Centerlines* centerlines)
{
  centerlines_ = centerlines;
}

void MouseCenterlineInteractorStyle::set_graphics(Graphics* graphics)
{
  graphics_ = graphics;
}

//------------
// OnKeyPress
//------------
//
void MouseCenterlineInteractorStyle::OnKeyPress() 
{
  // Get the keypress.
  vtkRenderWindowInteractor *rwi = this->Interactor;
  std::string key = rwi->GetKeySym();

  // Output the key that was pressed.
  std::cout << "[MouseCenterlineInteractorStyle::OnKeyPress] Pressed " << key << std::endl;

  // Undo. Removes the last slice.
  if (key == "s") {
    slice_mesh();

  } else if (key == "a") {
    graphics_->mesh_->extract_all_slices(graphics_->centerlines_->polydata_);

  // Quit.
  } else if ((key == "Escape") || (key == "q")) {
    exit(0);
  }

  // Forward event to catch VTK trackball keys (e.g. 'f').
  vtkInteractorStyleTrackballCamera::OnChar();
}

//-------------
// slice_mesh 
//-------------
//
void MouseCenterlineInteractorStyle::slice_mesh() 
{
  std::cout << "[slice_mesh] " << std::endl;

  // Get current screen location.
  int* clickPos = this->GetInteractor()->GetEventPosition();
  auto picker = vtkSmartPointer<vtkPropPicker>::New();
  picker->Pick(clickPos[0], clickPos[1], 0, this->GetDefaultRenderer());

  // Get selected actor and world point.
  if (picker->GetActor() == nullptr) {
    //vtkInteractorStyleTrackballCamera::OnLeftButtonDown();
    return;
  }
  double* pos = picker->GetPickPosition();
  std::cout << "[OnLeftButtonDown] Pick position (world coordinates) is: " << pos[0] << " " << pos[1] << " " << pos[2] << std::endl;

  // Create a sphere at the picked point.
  if (startSphere == nullptr) {
    startSphere = vtkSmartPointer<vtkSphereSource>::New();
    startSphere->SetThetaResolution(32);
    startSphere->SetPhiResolution(32);
    auto mapper = vtkSmartPointer<vtkPolyDataMapper>::New();
    mapper->SetInputConnection(startSphere->GetOutputPort());
    auto actor = vtkSmartPointer<vtkActor>::New();
    actor->GetProperty()->SetRepresentationToWireframe();
    actor->GetProperty()->SetRepresentationToPoints();
    actor->SetMapper(mapper);
    actor->GetProperty()->SetColor(0.0, 0.5, 0.5);
    actor->GetProperty()->SetOpacity(0.2);
    this->GetDefaultRenderer()->AddActor(actor);
  }

  // Get centerline data at the picked point.
  int index;
  double inscribedRadius;
  double normal[3]; 
  int cellID;
  centerlines_->locate_cell(pos, index, cellID, inscribedRadius, normal);

  startSphere->SetCenter(pos[0], pos[1], pos[2]);
  startSphere->SetRadius(inscribedRadius);

  double s = 0.5;
  double lpt[3] = { pos[0]+s*normal[0], pos[1]+s*normal[1], pos[2]+s*normal[2]};

  auto line = vtkLineSource::New();
  line->SetPoint1(pos);
  line->SetPoint2(lpt);
  line->Update();
  auto line_polydata = line->GetOutput();
  auto line_mapper = vtkPolyDataMapper::New();
  line_mapper->SetInputData(line_polydata);
  line_mapper->ScalarVisibilityOff();
  auto line_actor = vtkActor::New();
  line_actor->SetMapper(line_mapper);
  line_actor->GetProperty()->SetColor(1.0, 0.0, 1.0);
  line_actor->GetProperty()->SetLineWidth(4);
  this->GetDefaultRenderer()->AddActor(line_actor);

  // Slice the mesh.
  graphics_->mesh_->extract_slice(pos, inscribedRadius, normal);

  this->Interactor->GetRenderWindow()->Render();

}

vtkStandardNewMacro(MouseCenterlineInteractorStyle);
